# Varroa Counter
## 1.  Définition du problème

Les colonies d'abeilles du monde entier sont infestées par un parasite qui s'appelle le varroa destructor.
Ce parasite au nom barbare se fixe sur le corps des abeilles adultes et se nourrit de l'hémolymphe. Les femelles pénètrent aussi dans les cellules operculées pour se reproduire sur les larves, ce qui crée plusieurs générations au sein d'une même cellule.
Le varroa transmet des virus aux abeilles et affaiblit leur système immunitaire.
Si rien n'est entrepris pour stopper leur prolifération, les colonies finissent par s'effondrer durant l'automne ou l'hiver.


<img src="https://github.com/geda/varroa_detection/blob/main/varroa_destructor.jpg?raw=1" width="50%">

Une fois la récolte du miel effectuée, les apiculteurs effectuent différents traitements sur les colonies, notamment en utilisant de l'acide formique et de l'acide oxalique.
Une fois le traitement effectué, l'apiculteur dépose une planchette sous la ruche afin d'évaluer le degré d'infestation des colonies. Quelques jours après le traitement, les varroas morts tombent sur le fond de la ruche.
Les varroas ayant une taille de 1 à 2 mm, il devient très difficile de les compter lorsqu'ils sont nombreux. De plus, des résidus de cires d'abeilles tombent également des cadres, ce qui complique encore plus le comptage.

<img src="https://github.com/geda/varroa_detection/blob/main/fond_varroas.jpg?raw=1" width="50%">

Les varroas sont les petites formes sombres allongées et arrondies.

L'objectif de ce projet est d'estimer automatiquement le niveau d'infestation par les varroas à partir d'une image. Il ne s'agit pas d'obtenir un décompte exact, notamment lorsque les varroas sont peu nombreux, mais plutôt de fournir un ordre de grandeur fiable en cas d'infestation importante — par exemple, distinguer une image contenant 50 varroas d'une autre en contenant 200. Ce type d'évaluation, difficile et fastidieux à réaliser manuellement, est ainsi automatisé pour faciliter le travail de l'apiculteur.

## 2. Collecte de données
J'aimerais pouvoir simplement faire une photo de la planche complete et donné cette image assez large à un modèle pour l'inférence.
J'ai donc recherché des sets de données sur les varroas et j'en ai trouvé plusieurs disponibles sur https://universe.roboflow.com/. Malheureusement aucun dataset ne correspondait parfaitement à mes besoins:
* Images de varroas sur les abeilles et non sur la planche
* Images d'entraînement trop petites
* Images avec des varroas labellisés mais qui ne ressemblent pas vraiment à des varroas

J'ai donc créé un dataset avec mes propres images et j'ai uploadé le dataset sur roboflow sous le projet suivant: https://app.roboflow.com/varroa-counter/varroa-counter-large/2

### Inspection des données
Le dataset de données comprend des
* 19 images d'entraînement
* 4 images de validation
* 2 images de test

Au total 1723 varroas ont été labellisés. Plus de détails du dataset sont disponibles sous https://app.roboflow.com/varroa-counter/varroa-counter-large/health

## 3. Préparation des Données
J'ai labellisé les 32 images en identifiant plus 1000 varroas. J'ai effectué ce travail très chronophage directement sur roboflow.


Chaque image de ce set a donc maintenant un label associé. Un seul nom de classe est utilisé pour ce dataset: **varroa**
Les labels associés à une image sont sauvegardés au format YOLO et comprennent simplement une suite de nombres comme ceci:
* 0 0.02587890625 0.25439453125 0.0107421875 0.0166015625
* 0 0.021484375 0.27880859375 0.009765625 0.0166015625
* 0 0.0751953125 0.3349609375 0.009765625 0.015625
* ...

#### Explication du format YOLO

```
0 0.02587890625 0.25439453125 0.0107421875 0.0166015625
│      │              │            │            │
│      │              │            │            └── height (hauteur normalisée de la bounding box)
│      │              │            └── width (largeur normalisée de la bounding box)
│      │              └── y_center (position Y du centre, normalisée)
│      └── x_center (position X du centre, normalisée)
└── class_id (identifiant de la classe = 0 = varroa)
```

| Valeur | Signification | Exemple |
|--------|---------------|---------|
| `0` | ID de la classe | varroa (seule classe du dataset) |
| `0.0259` | x_center | Le centre est à 2.6% de la largeur de l'image (très à gauche) |
| `0.2544` | y_center | Le centre est à 25.4% de la hauteur de l'image |
| `0.0107` | width | La box fait 1.07% de la largeur de l'image |
| `0.0166` | height | La box fait 1.66% de la hauteur de l'image |

## 4. Analyse Exploratoire des Données (AED)
Les données contiennent des images avec des fonds de différentes couleurs et matières.

## 5. Feature Engineering
Le modèle devra faire de la détection d'objets. Un des challenges sera de détecter des objets très petits dans les images.
Les 2 features importantes de la détection d'objets seront:
* Classification d'image: déterminer si des varroas sont présents dans l'image
* Localisation d'objet: trouver la position des varroas à l'aide de _bounding boxes_

## 6. Modélisation
Je choisi la classification comme modèle d'apprentissage.
Etant donné que j'ai pas assez d'image pour entrainer un modèle depuis zéro, je recherche si je trouve un modèle existant.

J'ai trouvé le travail de recherche suivant: https://www.mdpi.com/2077-0472/15/9/969

**Référence:**
> Yániz, J.; Casalongue, M.; Martinez-de-Pison, F.J.; Silvestre, M.A.; Consortium, B.; Santolaria, P.; Divasón, J. *An AI-Based Open-Source Software for Varroa Mite Fall Analysis in Honeybee Colonies*. Agriculture 2025, 15, 969. https://doi.org/10.3390/agriculture15090969

Leurs modèle a été entraîné avec un dataset de 357 images sur plus de 500 epochs. Ils ont également livré un programme écrit en python permettant d'upload ses images et d'effectuer la détection des varroas avec leurs modèle.
Le code source est disponible sous github ainsi que leurs modèle entraîné: https://github.com/jodivaso/varrodetector/blob/main/model/weights/best.pt

En analysant le code j'ai trouvé particulièrement intéressant qu'ils utilisent une taille d'image assez grande de 6016 pixels (https://github.com/jodivaso/varrodetector/blob/main/varroa_mite_gui.py#L2507C42-L2507C46).

Je divise mon ensemble de donnée comme ceci:
* 26 images d'entraînement (80%)
* 4 images de validation (12%)
* 2 images de test (8%)

## 8. Evaluation du modèle
Je décide d'évaluer le modèle avec les données de mon dataset


In [ ]:
# checkout de mon repo
import os
import subprocess
import sys
from datetime import datetime
from pathlib import Path

# ===== CONFIGURATION - MODIFIE CES VALEURS =====
GITHUB_USERNAME = "geda"  # Ton username GitHub
REPO_NAME = "varroa_detection"       # Nom de ton repo
BRANCH = "main"                       # Branche Git
YOUR_EMAIL = "david@creasystem.ch"
YOUR_NAME = "David"

# Configuration de l'entraînement
EPOCHS = 50
IMG_SIZE = 640
BATCH_SIZE = 16
MODEL_SIZE = 'n'  # n, s, m, l, x (nano à extra-large)

# ===== FIN CONFIGURATION =====

print("🔧 Configuration...")

# Récupère le token depuis les secrets Colab
try:
    from google.colab import userdata
    GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
    print("✅ Token GitHub récupéré")
except Exception as e:
    print(f"❌ Erreur: Configure le secret GITHUB_TOKEN dans Colab")
    print(f"   Clique sur l'icône 🔑 à gauche et ajoute GITHUB_TOKEN")
    raise e

# URLs et chemins
REPO_URL = f"https://{GITHUB_USERNAME}:{GITHUB_TOKEN}@github.com/{GITHUB_USERNAME}/{REPO_NAME}.git"
REPO_PATH = f"/content/{REPO_NAME}"

def setup_repo():
    """Clone le repo et configure Git"""
    # Nettoie si existe déjà
    if os.path.exists(REPO_PATH):
        print(f"🧹 Nettoyage de {REPO_PATH}...")
        !rm -rf {REPO_PATH}

    # Clone le repo
    print(f"📥 Clone du repo {REPO_NAME}...")
    result = subprocess.run(
        ['git', 'clone', REPO_URL, REPO_PATH],
        capture_output=True,
        text=True,
        cwd='/content' # Explicitly set current working directory to /content
    )

    if result.returncode != 0:
        print(f"❌ Erreur lors du clone: {result.stderr}")
        raise Exception("Clone failed")

    # Change vers le repo
    os.chdir(REPO_PATH)

    # Configure Git
    subprocess.run(['git', 'config', 'user.email', YOUR_EMAIL], check=True)
    subprocess.run(['git', 'config', 'user.name', YOUR_NAME], check=True)

    print(f"✅ Repo cloné dans {REPO_PATH}")
    print(f"📁 Contenu du repo:")
    !ls -la

    return REPO_PATH

# Setup du repo
repo_path = setup_repo()
print(f"\n✅ Setup terminé!")
print(f"📍 Répertoire de travail: {os.getcwd()}")

🔧 Configuration...
✅ Token GitHub récupéré
📥 Clone du repo varroa_detection...
✅ Repo cloné dans /content/varroa_detection
📁 Contenu du repo:
total 916
drwxr-xr-x 5 root root   4096 Feb  8 07:36 .
drwxr-xr-x 1 root root   4096 Feb  8 07:36 ..
-rw-r--r-- 1 root root 178969 Feb  8 07:36 fond_varroas.jpg
drwxr-xr-x 8 root root   4096 Feb  8 07:36 .git
-rw-r--r-- 1 root root     24 Feb  8 07:36 .gitignore
-rw-r--r-- 1 root root  79560 Feb  8 07:36 IMG_0223_cropped_predict_16_zoom.jpg
drwxr-xr-x 3 root root   4096 Feb  8 07:36 model_mdpi_3291496
-rw-r--r-- 1 root root     18 Feb  8 07:36 README.md
-rw-r--r-- 1 root root  34768 Feb  8 07:36 train_varroa.ipynb
drwxr-xr-x 2 root root   4096 Feb  8 07:36 utils
-rw-r--r-- 1 root root 607799 Feb  8 07:36 varroa_destructor.jpg

✅ Setup terminé!
📍 Répertoire de travail: /content/varroa_detection


In [ ]:
# installation de roboflow et téléchargement de mon dataset
!pip install roboflow

import os
from roboflow import Roboflow

try:
    from google.colab import userdata
    api_key = userdata.get('ROBOFLOW_API_KEY')
except ImportError:
    api_key = os.getenv("ROBOFLOW_API_KEY")

rf = Roboflow(api_key=api_key)

# download du dataset depuis roboflow
project = rf.workspace("varroa-counter").project("varroa-counter-large")
version = project.version(3)
dataset = version.download("yolov11")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.8/91.8 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.8/66.8 kB 8.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.9/49.9 MB 52.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 60.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 103.5 MB/s eta 0:00:00
  Attempting uninstall: opencv-python-headless
    Found existing installation: opencv-python-headless 4.13.0.90
    Uninstalling opencv-python-headless-4.13.0.90:
      Successfully uninstalled opencv-python-headless-4.13.0.90
  Attempting uninstall: idna
    Found existing installation: idna 3.11
    Uninstalling idna-3.11:
      Successfully uninstalled idna-3.11
loading Roboflow workspace...
loading Roboflow project...



Extracting Dataset Version Zip to varroa-counter-large-3 in yolov11:: 100%|██████████| 62/62 [00:00<00:00, 789.93it/s]


In [ ]:
# Validation de mon modèle avec mes données de test
!pip install ultralytics
from ultralytics import YOLO

# Charger le modèle entraîné
model = YOLO('model_mdpi_3291496/weights/best.pt')

# Effectuer la détection sur l'image
results = model.val(
    data="varroa-counter-large-3/data.yaml",
    split='val',  # or 'val' for validation set
    imgsz=5712, # même valeur celle utilisée dans le code https://github.com/jodivaso/varrodetector/blob/main/varroa_mite_gui.py#L2507
    batch=4,
    conf=0.1,  # Seuil de confiance
    iou=0.5,
    max_det=2000,
    save_json=True,
    save=True,
    show_labels=False,
    show_conf=False,
    line_width=2
)

# Afficher les résultats
# Print results
print(f"Precision: {results.box.p}")
print(f"Recall: {results.box.r}")
print(f"mAP50: {results.box.map50}")
print(f"mAP50-95: {results.box.map}")

WARNING ⚠️ imgsz=[5712] must be multiple of max stride 32, updating to [5728]
Ultralytics 8.4.12 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (NVIDIA A100-SXM4-80GB, 81222MiB)
Model summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 4754.2±907.4 MB/s, size: 1907.1 KB)
val: Scanning /content/varroa_detection/varroa-counter-large-3/valid/labels.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 1.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 3.2s/it 3.2s
                   all          4         96       0.26       0.26      0.159     0.0605
Speed: 47.2ms preprocess, 29.3ms inference, 0.0ms loss, 1.9ms postprocess per image
Saving /content/varroa_detection/runs/detect/val18/predictions.json...
Results saved to /content/varroa_detection/runs/detect/val18
Precision: [    0.25972]
Recall: [    0.26042]
mAP50: 0.1587393967092202
m

Todo: commenter l'évaluation
# train
Precision: [    0.41507]
Recall: [    0.40815]
mAP50: 0.36199641493060963
mAP50-95: 0.09939633253557346

#valid
Precision: [    0.25972]
Recall: [    0.26042]
mAP50: 0.1587393967092202
mAP50-95: 0.06046614937409386

#test
Precision: [    0.13713]
Recall: [      0.375]
mAP50: 0.07988376105627809
mAP50-95: 0.02843958888244851

In [ ]:
# Validation de mon modèle avec mes données de test
from ultralytics import YOLO

# Charger le modèle entraîné
model = YOLO('model_mdpi_3291496/weights/best.pt')

results = model.train(data="varroa-counter-large-3/data.yaml",
                      epochs=1,
                      batch=4,
                      imgsz=4000,
                      freeze=10,
                      max_det=2000,
                      conf=0.1,
                      iou = 0.5
                      )

Ultralytics 8.4.12 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (NVIDIA A100-SXM4-80GB, 81222MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=0.1, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=varroa-counter-large-3/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=1, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=10, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=5712, int8=False, iou=0.5, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=2000, mixup=0.0, mode=train, model=model_mdpi_3291496/weights/best.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=train2, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overla

In [ ]:
import subprocess
import os
from datetime import datetime
from google.colab import userdata

# Configure Git (these variables are already set in f247fe02, but repeated here for robustness in isolated execution)
# However, it's generally better to ensure f247fe02 runs first.
# Assuming REPO_PATH, GITHUB_USERNAME, etc. are defined from f247fe02.
# If you run this cell independently, ensure these variables are in your environment.
GITHUB_USERNAME = "geda"  # Ton username GitHub
REPO_NAME = "varroa_detection"       # Nom de ton repo
BRANCH = "main"                       # Branche Git
YOUR_EMAIL = "david@creasystem.ch"
YOUR_NAME = "David"

# URLs et chemins
REPO_PATH = f"/content/{REPO_NAME}"

# Change vers le repo (assuming REPO_PATH is defined from f247fe02)
os.chdir(REPO_PATH)

# Configure Git
subprocess.run(['git', 'config', 'user.email', YOUR_EMAIL], check=True)
subprocess.run(['git', 'config', 'user.name', YOUR_NAME], check=True)

# Ajoute les fichiers et répertoires pertinents
print(f"Adding 'train_varroa.ipynb' from {os.path.join(REPO_PATH, 'train_varroa.ipynb')}")
subprocess.run(['git', 'add', 'train_varroa.ipynb'], check=False) # check=False because it might not exist if it's a fresh clone

print(f"Adding 'runs/' directory from {os.path.join(REPO_PATH, 'runs/')}")
subprocess.run(['git', 'add', 'runs/'], check=False) # check=False because runs/ might not have been created yet, or might be empty

# Check if there are any staged changes
staged_changes_exist = False
try:
    # git diff --cached --exit-code returns 0 if no differences (nothing staged), 1 if differences
    # We expect a non-zero exit code if changes exist, so check=True will raise an exception.
    subprocess.run(['git', 'diff', '--cached', '--exit-code'], check=True, capture_output=True)
    print("ℹ️ Aucun changement stagé à commiter.")
    staged_changes_exist = False
except subprocess.CalledProcessError as e:
    # If we reach here, it means git diff --cached --exit-code returned a non-zero exit code.
    # For this specific command, exit code 1 means there ARE staged changes.
    if e.returncode == 1:
        staged_changes_exist = True
        print("✓ Changements stagés détectés.")
    else:
        # Re-raise for unexpected errors from git diff --cached --exit-code
        print(f"❌ Erreur inattendue lors de la vérification des changements stagés: {e.stderr}")
        raise e

if staged_changes_exist:
    # Commit
    message = f"Training - {datetime.now().strftime('%Y-%m-%d %H:%M')}"
    subprocess.run(['git', 'commit', '-m', message], check=True)
    print(f"✅ Commit créé : {message}")

    # Push
    GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
    subprocess.run(
        ['git', 'push',
         f"https://{GITHUB_USERNAME}:{GITHUB_TOKEN}@github.com/{GITHUB_USERNAME}/{REPO_NAME}.git",
         BRANCH],
        check=True
    )

    print("✅ Résultats sur GitHub!")
else:
    print("Pas de commit ou push car aucun changement stagé n'a été détecté.")

Adding 'train_varroa.ipynb' from /content/varroa_detection/train_varroa.ipynb
Adding 'runs/' directory from /content/varroa_detection/runs/
✓ Changements stagés détectés.
✅ Commit créé : Training - 2026-02-08 07:37
✅ Résultats sur GitHub!
